Add docs for implementing a server (#808)
connectrpc/conformance: docs/testing_servers.md

# Testing Servers

Testing servers with the conformance runner involves the following steps:

* Implementing the ConformanceService endpoints.
* Making an executable file that will read messages from `stdin` and write messages to `stdout`.

> * Defining any configuration for what your server supports. For more information on how to do this, see the [Configuration Files](./configuring_and_running_tests.md#configuration-files) section in the [Configuring and Running Tests documentation](./configuring_and_running_tests.md).

Reviewer: I don't think we need two links here. How about this?
```suggestion
* Defining any configuration for what your server supports. For more information on how to do this, see the docs for [configuring and running tests](./configuring_and_running_tests.md#configuration-files).
```

## Process

The basic process for working with servers in the conformance suite will work as follows:

> * The conformance runner will pass a `ServerCompatRequest` message via `stdin`. This message will contain all the details necessary

Reviewer: We need a little more clarity on what it means to "pass a message via stdin". In particular, this should describe how the program will need to read four bytes for the 32-bit (network byte order) size of the message, then read that many bytes, and then de-serialize those bytes into this message type.

Same for the "write a message to stdout" below.

  for your implementation to start its server-under-test. Note that the request does not specify a port to listen on.
  Implementations should instead pick an available ephemeral port according to their OS and return that value in the response.

  Fields in the request are:

>   * `protocol` specifies what protocol will be used to run the tests. 

Reviewer: If the server implementation under test always supports multiple protocols, then it is fine to ignore this field. But this is presented in case the server needs to do anything extra to support particular protocols.

>   * `http_version` specifies which HTTP version will be used.

Reviewer: Maybe we could clarify here, too. Some HTTP server libraries may automatically provide support for both HTTP 1.1 and HTTP/2, in which case it is fine for the server program to always support both and ignore this field.

So both of these fields really mean "support _at least_ this protocol and version; fine to also support others".

>   * `use_tls` specifies whether your server should generate a self-signed certificate. Clients will be configured to trust

Reviewer: FWIW, it is not _required_ for the cert to be self-signed. It can be any TLS certificate where the subject matches the host name that gets sent back in the `ServerCompactResponse`.

But we explicitly _allow_ self-signed (and "localhost" as the subject) to make it easier to implement the server w/out having to worry about certificate issuance, managing/distributing certificate secrets, etc.

     this certificate when connecting. If `true`, the generated certificate should be returned in the `pem_cert` field of the
     `ServerCompatResponse`. If this is set to `false`, the server should not use TLS and instead use a plaintext/unencrypted socket.
  * `client_tls_cert` represents a PEM-encoded certificate that, if provided, clients will use to authenticate themselves. 
     If specified, servers should require that clients provide certificates and they should ensure the presented certificate is valid.
     Note that this will always be empty if `use_tls` is `false`.
  * `message_receive_limit` specifies the maximum size in bytes for a message. If this value is non-zero, servers should reject
     any message from a client that is larger than the size indicated.

* Once the server is started, your implementation should then write a `ServerCompatResponse` message to `stdout`. This will
  provide the conformance runner with details about your running server.

>   * `host` should be set with the host where your server is running. 

Reviewer: This should usually be "127.0.0.1", unless the program actually starts a remote server to which the client should connect.

  * `port` should be set with the port number where your server is listening.

>   * `pem_cert` should contain the self-signed certificate generated by your server if `use_tls` was set to `true`. Clients

Reviewer: ```suggestion
  * `pem_cert` should contain the self-signed certificate generated by your server if `use_tls` was set to `true`, in PEM format. Clients
```

* Any response data
* Information observed from the request such as:
  * Request headers
  * Any timeout specified
  * All request bodies received
  * Any query parameters observed (only applicable for GET operations such as `IdempotentUnary`)

In all, there are six total endpoints you will need to implement. Below is a brief description of each with helpful pseudocode
for your implementation.

### Unary

The `Unary` endpoint is a unary operation that accepts a single request of type `UnaryRequest` and returns a single 
response of type `UnaryResponse`. 

The `UnaryRequest` contains a `response_definition` that tells the server how to respond. This definition contains a 
`oneof` which specifies whether the server should return valid response data or return an error. Servers should also 
allow this response definition to be unset. In which case, they should set no response headers or trailers and return 
no response data. The returned conformance payload should only contain the observed request information.

**Pseudocode**

```text
read the request

capture any request headers and the actual request body as request info

if a response definition was specified then

>   if response definition specifies valid response data then

Reviewer: Maybe we can tweak this condition and then swap the two branches? I think it would be more clear if it were more like "if response definition specifies an error then"

    build a conformance payload

    set the request info and response data into the conformance payload

  else

>     build an error with the specified error and set the request info into the error details

Reviewer: I think we might want a sentence or paragraph (outside the pseudo-code) about error details and storing the request info as an error detail -- and reference the precise proto message type (and even link to its docs in the BSR).

  set any response headers or trailers indicated in the response definition

> else

Reviewer: I think the conditionals could be simplified to something like:
```
if response definition specifies an error then

   // do error

else

    build a conformance payload with the request info
    if a response definition exists then set the payload's data

end if
```
(For that last line: I wonder if it would be easier to read with explicit "end" lines or even curly braces. As is, it could be a little confusing for anyone but a Python developer 😛. WDYT?)

  build a conformance payload

  set the request info into the conformance payload

sleep for any specified response delay

> return the response with conformance payload or the error depending on which was specified

Reviewer: This probably depends on the programming language. In some languages, errors are thrown/raised instead of returned. Maybe tweak the verb?
```suggestion
return the response with conformance payload or raise the error, depending on which was specified
```

```

For the full documentation on implementing the `Unary` endpoint, click [here][unary].

### IdempotentUnary

The `IdempotentUnary` endpoint is also a unary operation. It accepts a request of type  `IdempotentUnaryRequest` and returns
a single response of type `IdempotentUnaryResponse`. It should be handled in mostly the same way as `Unary`.

> The only major difference is that this endpoint should be invoked via an HTTP `GET`. As a result, there is no request body

Reviewer: I think it's worth clarifying this a little more. In particular that implementations that only support gRPC or gRPC-Web do not need to implement it. Also, Connect implementations that do not support GET requests also don't need to implement it.

so the endpoint should read any query params and set them accordingly in the `connect_get_info` field of the 
returned `ConformancePayload`. This RPC is the only one that sets the `connect_get_info` field.

For the full documentation on implementing the `IdempotentUnary` endpoint, click [here][idempotentunary].

### Unimplemented

The `Unimplemented` endpoint is also a unary operation, but contrary to the above unary endpoints, the implementation

> of `Unimplemented` should simply return an `unimplemented` error. It is not necessary to echo back any request information or

Reviewer: Actually, it should not be implemented at all. This tests that the server framework can automatically return the "unimplemented" error for unimplemented methods. Part of the gRPC spec is that adding methods to a service is a safe/compatible change, and that server frameworks should handle this by automatically responding to methods that are missing from the actual server handler implementation. (In connect-go, this actually gets handled by the generated code, in the `Unimplemented...` generated server handler struct, instead of elsewhere in the connect-go runtime code.)

  then alternate, reading another request and then sending back another response, etc.
  
* if the server receives a request and has no responses to send, it
  should throw the error specified in the request.
  
* the service should echo back all request properties in the first response
  including the last received request. Subsequent responses should only
  echo back the last received request.
  
* if the `response_delay_ms` duration is specified, the server should wait the given
  duration after reading the request before sending the corresponding
  response.
  
If the `full_duplex` field is false:

* the handler should read all requests until the client is done sending.
  Once all requests are read, the server should then send back any responses
  specified in the response definition.
  
* the server should echo back all request properties, including all request
  messages in the order they were received, in the first response. Subsequent
  responses should only include the message data in the data field.
  
* if the `response_delay_ms` duration is specified, the server should wait that
  long in between sending each response message.

**Pseudocode**

> ```text

Reviewer: I think this might be easier to follow if we split it into two -- pseudo-code for the full-duplex logic and separate pseudo-code for the half-duplex logic.
